src/EnergyPlus/HVACInterfaceManager.cc

```diff
@@ -212,12 +218,42 @@ void UpdateHVACInterface(EnergyPlusData &state,
             OutOfToleranceFlag = true; // Something has changed--resimulate the other side of the loop
         }
 
+        if (state.dataContaminantBalance->Contaminant.CO2Simulation) {
+            TmpRealARR = airLoopConv.HVACCO2DemandToSupplyTolValue;
+            airLoopConv.HVACCO2DemandToSupplyTolValue[0] = std::abs(state.dataLoopNodes->Node(OutletNode).CO2 - thisInletNode.CO2);
+            for (int logIndex = 1; logIndex < DataConvergParams::ConvergLogStackDepth; logIndex++) {
+                airLoopConv.HVACCO2DemandToSupplyTolValue[logIndex] = TmpRealARR[logIndex - 1];
+            }
+            if (airLoopConv.HVACCO2DemandToSupplyTolValue[0] > DataConvergParams::HVACCO2Toler) {
+                airLoopConv.HVACCO2NotConverged[0] = true;
+                OutOfToleranceFlag = true; // Something has changed--resimulate the other side of the loop
+            }
+        }
+
+        if (state.dataContaminantBalance->Contaminant.GenericContamSimulation) {
+            TmpRealARR = airLoopConv.HVACGenContamDemandToSupplyTolValue;
+            airLoopConv.HVACGenContamDemandToSupplyTolValue[0] = std::abs(state.dataLoopNodes->Node(OutletNode).GenContam - thisInletNode.GenContam);
+            for (int logIndex = 1; logIndex < DataConvergParams::ConvergLogStackDepth; logIndex++) {
+                airLoopConv.HVACGenContamDemandToSupplyTolValue[logIndex] = TmpRealARR[logIndex - 1];
+            }
+            if (airLoopConv.HVACGenContamDemandToSupplyTolValue[0] > DataConvergParams::HVACGenContamToler) {
+                airLoopConv.HVACGenContamNotConverged[0] = true;
+                OutOfToleranceFlag = true; // Something has changed--resimulate the other side of the loop
+            }
+        }
+
     } else if (CalledFrom == DataConvergParams::CalledFrom::AirSystemSupplySideDeck1) {
 
         airLoopConv.HVACMassFlowNotConverged[1] = false;
         airLoopConv.HVACHumRatNotConverged[1] = false;
         airLoopConv.HVACTempNotConverged[1] = false;
         airLoopConv.HVACEnergyNotConverged[1] = false;
+        if (state.dataContaminantBalance->Contaminant.CO2Simulation) {
+            airLoopConv.HVACCO2NotConverged[1] = false;
+        }
+        if (state.dataContaminantBalance->Contaminant.GenericContamSimulation) {
+            airLoopConv.HVACGenContamNotConverged[1] = false;
+        }
```

Comment:
Oof, I really don't like further propagating these numbers.  As far as I can tell, the index in this array is based on where we are calling from?  Is that actually how it works?  I'm not saying we need to go in and gut things for this fix, but it's just something that I think should be cleaned up.  At a minimum, shouldn't we at least have a constexpr variable for "0", "1", and "2"?  And more importantly, does it make sense for these to be arrays at all?  Feels more like we should have a little struct and just create 3 different variable instances of that struct.  Rant over.

Reply:
@Myoldmopar Your understanding is correct. The index in the array represents a call-from type as DataConvergParams::CalledFrom. I also agree to have a constexpr variable for "0", "1", and "2" with clear representation. Since constexpr variable change is beyond the current issue scope, I would like to suggest to create a new issue to handle a new constexpr variable.